packages/contracts-periphery/contracts/universal/citizen-house/Badge.sol

```diff
@@ -0,0 +1,110 @@
+        _;
+    }
+
+    address public AdminContract;
+    string private baseURI;
+    uint256 private totalSupply;
+
+    /*///////////////////////////////////////////////////////////////
+                            CONSTRUCTOR
+  //////////////////////////////////////////////////////////////*/
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _baseURI
+    ) payable ERC721(_name, _symbol) {
+        baseURI = _baseURI;
+    }
+
+    /// @notice Mint
+    /// @dev Mints the soulbound ERC721 token.
+    function mint(address _citizen) external onlyAdmin {
+        require(
+            AdminContract != address(0),
+            "Only Admin Contract can mint & Admin Contract not set"
+        );
+        _mint(_citizen, totalSupply++);
+    }
+
+    /// @notice Burn
+    /// @dev Burns the soulbound ERC721.
+    /// @param _id The token URI.
+    function burn(uint256 _id) external {
+        require(_ownerOf[_id] == msg.sender, "Error: Sender is not owner");
+        _burn(_id);
+    }
+
+    /// @notice Token URI
+    /// @dev Generate a token URI.
+    /// @param _id The token URI.
+    function tokenURI(uint256 _id) public view override returns (string memory) {
+        return string(abi.encodePacked(baseURI, _id));
+    }
+
+    /// @notice AdminContract
+    /// @dev Update Admin Contract
+    function updateAdminContract(address _adminContract) external onlyOwner {
+        AdminContract = _adminContract;
+    }
+
+    /// @notice Withdraw
+    /// @dev Withdraw the contract ETH balance
+    function withdraw() external onlyOwner {
+        SafeTransferLib.safeTransferETH(msg.sender, address(this).balance);
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            Make the Badge Soul Bound
+  //////////////////////////////////////////////////////////////*/
+
+    /// @notice Transfer ERC721
+    /// @dev Override the ERC721 transferFrom method to revert
+    function transferFrom(
+        address,
+        address,
+        uint256
+    ) public pure override {
+        // Make it ~*~ Soulbound ~*~
+        revert Soulbound("transferFrom(address, address, uint256)");
+    }
+
+    /// @notice Approve ERC721
+    /// @dev Override the ERC721 Approve method to revert
+    function approve(address, uint256) public pure override {
+        revert Soulbound("approve(address, uint256)");
+    }
+
+    /// @notice setApprovalForAll ERC721
+    /// @dev Override the ERC721 setApprovalForAll method to revert
+    function setApprovalForAll(address, bool) public pure override {
+        revert Soulbound("setApprovalForAll(address, uint256)");
+    }
+
+    // https://eips.ethereum.org/EIPS/eip-165#simple-summary
+    function supportsInterface(bytes4 _interfaceId) public pure override(ERC721) returns (bool) {
+        return
+            _interfaceId == 0x7f5828d0 || // ERC165 Interface ID for ERC173
```

Comment:
Can you use a style similar to this? https://github.com/ethereum-optimism/optimism/blob/fa9823f373256a4ce3e27aee61bfcd340a316784/packages/contracts-bedrock/contracts/universal/OptimismMintableERC20.sol#L90